### PR TITLE
Feature/chat interaction

### PR DESCRIPTION
**Pull request by greegouze**
## Description

// Create feature for starting a new conversation with the interacting user,
//Creation of a conversation when two users interact with each other

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Area of change

- [x] Frontend
- [x] Backend
- [ ] Other: .....

## General checklist:

- [x] My code follows the style guidelines of this project
- [x] I ran `bun run format`/`bun run lint` before commit
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation where needed
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings
- [x] I have updated `develop` and merged to my branch before submitting pull request
- [x] My pull request generate no conflicts with `develop` branch
- [x] I requested code review from other team members

## Frontend checklist:

- [x] I followed guidelines for `HTML`/`JSX`/`Typescript`
- [x] My `Javascript` generate no new console errors
- [x] I tested my code cross browsers
- [ ] My slice is pixel perfect for both desktop and mobile according to design
- [ ] I conducted basic QA to assure all features are working
- [x] I tested responsive for mobile and tablet resolutions

## Backend checklist:

- [x] I tested admin by manually adding content from zero
- [x] I created easy to use admin experience which is self-explanatory
- [x] I added description to admin fields in hard-to-understand areas
- [x] I followed guidelines for naming `Typescript` variables
- [ ] I conducted basic QA to assure all features are working

### CODE REVIEW

File: packages/backend/api/src/interaction/interaction-crud.ts [hunk 1]
```diff
@@ -5,13 +5,19 @@ import { jsonObjectFrom } from 'kysely/helpers/mysql';
 import { db } from '@app/backend-shared';
 import { type ActionBody, actionSchema, receiverSchema } from '@app/shared';
 import type { Request as ExpressRequest } from '@app/shared';
+import type { Conversation } from '@app/shared';
```
Comment (Mikazian) [inline packages/backend/api/src/interaction/interaction-crud.ts:8]: Peut être tous sur la même ligne

File: packages/backend/api/src/interaction/interaction-crud.ts [hunk 2]
```diff
@@ -5,13 +5,19 @@ import { jsonObjectFrom } from 'kysely/helpers/mysql';
...
 
 import { getUserId } from '@/middlewares/auth-handlers';
 import { getSuperLikeCount } from '@/middlewares/interaction-handlers';
+import { verifyInteractions } from '@/middlewares/verify-match-handlers';
 
+type NewConversation = Omit<Conversation, 'id'>;
```
Comment (Mikazian) [inline packages/backend/api/src/interaction/interaction-crud.ts:14]: Fait plutôt le typage côté shared puis tu importes ce type ici au lieu de le créer ici

File: packages/backend/api/src/user/message.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+messageRouter.get(
+  '/:userId/conversations',
+  getUserId,
+  async (req: Request, res) => {
+    try {
+      const userId = req.userId as number;
+
+      //Retrieves all conversations of a user
+      const conversation = await db
+
```
Comment (Mikazian) [inline packages/backend/api/src/user/message.ts:20]: ta un retour à la ligne ici

File: packages/backend/api/src/user/message.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+//Route to retrieve the conversation with my interlocutor
+messageRouter.get(
+  '/:userId/conversations/:conversationId',
+  getUserId,
+  async (req: Request, res) => {
+    try {
+      const conversationId = Number.parseInt(req.params.conversationId);
+
+      const { userId } = req.params;
```
Comment (Mikazian) [inline packages/backend/api/src/user/message.ts:79]: Number.parseInt

File: packages/backend/api/src/user/message.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      const userIniator = req.userId as number;
+
+      if (Number(userId) !== userIniator) {
+        return res.status(403).send('Unauthorized');
+      }
+
+      const messages = await db
+
```
Comment (Mikazian) [inline packages/backend/api/src/user/message.ts:87]: retirer le retour à la ligne

File: packages/frontend/web/src/components/message/Conversation.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import { useConversation } from '@/contexts/ConversationContext';
+
+import CrossIcon from '../icons/CrossIcon';
+import SendIcon from '../icons/SendIcon';
+import BulletConversation from './BulletConversation';
+
+export default function Conversation() {
+  const { userId, conversation, setIsVisible, error, selectedConversation } =
+    useConversation();
+  const { register, handleSubmit, reset } = useForm();
```
Comment (Mikazian) [inline packages/frontend/web/src/components/message/Conversation.tsx:16]: zodResolver ?

File: packages/frontend/web/src/components/message/ConversationsList.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                  >
+                    <div className='flex w-full items-center gap-2 overflow-hidden'>
+                      <div className='h-14 w-14 shrink-0 overflow-hidden rounded-full'>
+                        <img
+                          src={
+                            userId === conversation.user_1.id
+                              ? conversation.user_2.picture_path
+                              : conversation.user_1.picture_path
+                          }
+                          // alt={`Picture of ${conversation.receiver[0].receiver_name}`}
```
Comment (Mikazian) [inline packages/frontend/web/src/components/message/ConversationsList.tsx:84]: code commenté

File: packages/frontend/web/src/components/message/ConversationsList.tsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                        </div>
+                      </div>
+                    </div>
+                    <div className='flex w-24 items-end justify-end '>
+                      <div className='self shrink-0 text-right text-sm'>
+                        <DateComponent
+                          date={
+                            Boolean(conversation.messages)
+                              ? conversation.messages.sent_at
+                              : ''
```
Comment (Mikazian) [inline packages/frontend/web/src/components/message/ConversationsList.tsx:108]: undefined

File: packages/frontend/web/src/contexts/ConversationContext.tsx [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+type Messages = Omit<Message, 'sender_id'> & {
+  sender_name: string;
+};
+
+interface Conversation {
+  conversation_id: number;
+  user_1: User;
+  user_2: User;
+  messages: Messages[];
+}
```
Comment (Mikazian) [inline packages/frontend/web/src/contexts/ConversationContext.tsx:50]: beaucoup de typage ici, mettons plutôt la logique du typage dans le dossier shared

File: packages/frontend/web/src/contexts/ConversationContext.tsx [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  const [error, setError] = useState<string>();
+
+  const [conversationId, setConversationId] = useState<number>();
+
+  const [isVisible, setIsVisible] = useState<boolean>(false);
+
+  const { conversationsList, messagesCount, fetchConversations } =
+    useAllConversations();
+
+  /* const navigate = useNavigate();  */
```
Comment (Mikazian) [inline packages/frontend/web/src/contexts/ConversationContext.tsx:92]: code commenté

File: packages/frontend/web/src/contexts/ConversationContext.tsx [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    if (Boolean(conversationId) && conversationsList !== undefined) {
+      const controller = new AbortController();
+
+      const signal = controller.signal;
+      const fetchMessage = async () => {
+        const response = await fetch(
+          `/api/users/${userId}/conversations/${conversationId}`,
+          {
+            signal,
+            credentials: 'include',
```
Comment (Mikazian) [inline packages/frontend/web/src/contexts/ConversationContext.tsx:141]: enlever le cookie

File: packages/frontend/web/src/hooks/use-all-conversations.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    [userId],
+  );
+
+  // Fetch conversations between the user logged in and other users
+  useEffect(() => {
+    const controller = new AbortController();
+    const signal = controller.signal;
+
+    fetchConversations({ signal }).catch((error) => {
+      throw new Error(`Failed to fetch messages: ${String(error)}`);
```
Comment (Mikazian) [inline packages/frontend/web/src/hooks/use-all-conversations.ts:48]: enlever le throw et créer un state error pour setError

File: packages/frontend/web/src/hooks/use-interactions.ts [hunk 13]
```diff
@@ -76,14 +79,26 @@ export default function useInteractions({ ...props }) {
...
       });
 
+      const fetchInteractionsVerify = async () => {
+        await fetch(`/api/users/interactions/verify`, {
+          signal,
+        });
+      };
+
+      await fetchInteractionsVerify().catch((error) => {
+        throw new Error(`Fail to fetch interactions verify: ${String(error)}`);
```
Comment (Mikazian) [inline packages/frontend/web/src/hooks/use-interactions.ts:97]: pareil enlever le throw

File: packages/frontend/web/src/pages/Home.tsx [hunk 14]
```diff
@@ -19,10 +21,24 @@ export default function Home() {
...
-  const hasCompleteProfile =
-    userProfile &&
+  const hasCompleteProfile = userProfile; /*  &&
+    (Boolean(userProfile.name) ||
+      Boolean(userProfile.birthdate) ||
+      Boolean(userProfile.gender) ||
+      Boolean(userProfile.city) ||
+      userProfile.languages.length > 0 ||
+      userProfile.technologies.length > 0 ||
+      userProfile.hobbies.length > 0) */
```
Comment (Mikazian) [inline packages/frontend/web/src/pages/Home.tsx:38]: Code commenté

**Review by Mikazian — DISMISSED** [review]

**Review by YDrogen — CHANGES_REQUESTED** [review]

**Review by Mikazian — APPROVED** [review]

**Review by YDrogen — APPROVED** [review]

**Pull request merged**